clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreAwarePartitionWiseKafkaConsumerService.java

```diff
@@ -0,0 +1,196 @@
+      final AggKafkaConsumerServiceStats stats,
+      final boolean isKafkaConsumerOffsetCollectionEnabled,
+      final ReadOnlyStoreRepository metadataRepository,
+      final boolean isUnregisterMetricForDeletedStoreEnabled) {
+    super(
+        poolType,
+        consumerFactory,
+        consumerProperties,
+        readCycleDelayMs,
+        numOfConsumersPerKafkaCluster,
+        ingestionThrottler,
+        kafkaClusterBasedRecordThrottler,
+        metricsRepository,
+        kafkaClusterAlias,
+        sharedConsumerNonExistingTopicCleanupDelayMS,
+        topicExistenceChecker,
+        liveConfigBasedKafkaThrottlingEnabled,
+        pubSubDeserializer,
+        time,
+        stats,
+        isKafkaConsumerOffsetCollectionEnabled,
+        metadataRepository,
+        isUnregisterMetricForDeletedStoreEnabled,
+        StoreAwarePartitionWiseKafkaConsumerService.class.toString());
+    this.storeRepository = metadataRepository;
+  }
+
+  @Override
+  protected synchronized SharedKafkaConsumer pickConsumerForPartition(
+      PubSubTopic versionTopic,
+      PubSubTopicPartition topicPartition) {
+    String resourceIdentifier = getResourceIdentifier(versionTopic);
+    int minLoad = Integer.MAX_VALUE;
+    SharedKafkaConsumer minLoadConsumer = null;
+    Map<Integer, Integer> consumerToLoadMap = new VeniceConcurrentHashMap<>();
+    for (SharedKafkaConsumer consumer: getConsumerToConsumptionTask().keySet()) {
+      int index = getConsumerToConsumptionTask().indexOf(consumer);
+      if (topicPartition.getPubSubTopic().isRealTime()
+          && alreadySubscribedRealtimeTopicPartition(consumer, topicPartition)) {
+        getLOGGER().info(
+            "Consumer id: {} has already subscribed the same real time topic-partition: {}, will assign MAX load to avoid being picked.",
+            index,
+            topicPartition);
+        continue;
+      }
+      int baseAssignmentCount = consumer.getAssignmentSize();
+      int storeSubscriptionCount =
+          getResourceIdentifierToConsumerMap().getOrDefault(resourceIdentifier, Collections.emptyMap())
+              .getOrDefault(consumer, 0);
+      int overallLoad = storeSubscriptionCount * IMPOSSIBLE_MAX_PARTITION_COUNT_PER_CONSUMER + baseAssignmentCount;
+      if (overallLoad < minLoad) {
+        minLoadConsumer = consumer;
+        minLoad = overallLoad;
+      }
+      consumerToLoadMap.put(index, overallLoad);
+    }
+    if (minLoad == Integer.MAX_VALUE) {
+      throw new IllegalStateException(
+          "Unable to find least loaded consumer entry. Current load map: " + consumerToLoadMap);
+    }
+    // Update resource identifier to consumer load map.
+    Map<PubSubConsumerAdapter, Integer> consumerMap = getResourceIdentifierToConsumerMap()
+        .computeIfAbsent(resourceIdentifier, key -> new VeniceConcurrentHashMap<>());
+    int existCount = consumerMap.getOrDefault(minLoadConsumer, 0);
+    consumerMap.put(minLoadConsumer, existCount + 1);
```

Comment:
I am wondering what the main reason is to maintain this map considering the above logic will count it every time when picking up the consumer.
Can we remove this map completely?

Reply:
Hmm, the above calculation reads the existing counting from the map. Are you suggesting, instead of maintaining a map, everytime we just iterate through the set of the assignment and calculate it ad-hoc?

Reply:
I am actually thinking that - if we simplify the problem to always use store name as identifier regardless of hybrid store or not, then we can remove this map, as each check will only extract store name from assignment. Otherwise, we will have to always refer to store repo for every assignment to compute the identifier, if we don't keep this map.

Reply:
Got it.
I am fine to pursue the simplified approach as we do have plan to unsubscribe batch-only store after completion and make it as the default setup.
Also batch-only store won't show up in AA/WC leader consumer pool, which is the main concern.